Feat/custom rate limit policies (#1252)
auth0/docs-v2: main/docs/troubleshoot/customer-support/operational-policies/rate-limit-policy/custom-rate-limit-policies.mdx

---
title: Custom Rate Limit Policies
description: Set per-application rate limit ceilings to protect your tenant from misconfigured or misbehaving applications.
---

> Custom rate limit policies let you set per-application request ceilings on your tenant's Authentication API. When an application calls Auth0 (requesting tokens, starting authorization flows, or refreshing credentials) each request counts toward that application's configured limit. If an application exceeds its ceiling due to a bug, retry loop, or unexpected behavior, Auth0 throttles it before it can consume your tenant's shared rate limit capacity.

Reviewer: ```suggestion
Custom rate limit policies let you set per-application request ceilings on your tenant's calls to [Authentication API](https://auth0.com/docs/api/authentication). When an application calls Auth0 (requesting tokens, starting authorization flows, or refreshing credentials) each request counts toward that application's configured limit. If an application exceeds its ceiling due to a bug, retry loop, or unexpected behavior, Auth0 throttles it before it can consume your tenant's shared rate limit capacity.
```

## How policy evaluation works

Auth0 evaluates custom rate limit policies in a strict hierarchy, from most specific to most general. The first matching level is enforced and all broader levels are skipped.

1. **Client ID**: A policy targeting a specific application by its Client ID. If a matching policy exists, evaluation stops here.
2. **Group**: Grouped profiles (such as third-party applications or CIMD) that apply a single shared limit to the *combined* traffic from all applications in the group. All applications in the group draw from the same policy. There is no per-application subdivision within the group. If the application belongs to a group with a policy, evaluation stops here.
3. **Global / Default**: A fallback policy that applies to any application not covered by a Client ID or Group policy. Each application gets its own independent policy at the configured limit. The limit is shared, but the policies are not. This means each application is evaluated separately against the same ceiling.

This hierarchy protects against both individual misbehaving applications and aggregate surges from entire classes of applications.

### Pooled counters (Group policies)

> Group policies use a single shared counter for all applications in the group. Every request from any group member draws from the same bucket.

Reviewer: ```suggestion
Group policies use a single, shared counter for all applications in the group. Every request from any group member draws from the same bucket.
```

> For example, if you create a Group policy called "Third-Party Clients" with a limit of 100 RPS, and the group contains five applications, the 100 RPS limit applies to their combined traffic. If one application sends 90 RPS and the others send 5 RPS each, the group has hit its limit (90 + 5 + 5 + 5 + 5 = 110 RPS) and subsequent requests from any application in the group are throttled. There is no guaranteed allocation per application within the group.

Reviewer: ```suggestion
For example, if you create a Group policy called "Third-Party Clients" with a limit of 100 RPS, and the group contains five applications, the 100 RPS limit applies to their combined traffic. If one application sends 90 RPS and the others send 5 RPS each, the group hits its limit (90 + 5 + 5 + 5 + 5 = 110 RPS). Auth0 throttles subsequent requests from any application in the group. There is no guaranteed allocation per application within the group.
```

If one third-party application needs different treatment, assign it a Client ID policy, which takes precedence over the group limit.

#### Third-party apps vs. CIMD clients

Although CIMD (Client-Initiated Managed Delegation) clients are technically third-party applications, Auth0 places them in a separate group from traditional third-party apps. This means two distinct group buckets exist:

> - **Third-party apps** (Client ID prefix: `tpa_`) — Traditional third-party applications, including those created via the Management API or Dynamic Client Registration (DCR).

Reviewer: ```suggestion
- **Third-party apps** (Client ID prefix: `tpa_`):  Traditional third-party applications, including those created via the Management API or Dynamic Client Registration (DCR).
```

Reviewer: For existing convention

> - **CIMD clients** (Client ID prefix: `https://`) — Applications created through Client-Initiated Managed Delegation.

Reviewer: ```suggestion
- **CIMD clients** (Client ID prefix: `https://`): Applications created through Client-Initiated Managed Delegation.
```

| --- | --- | --- |
| Token requests | `/oauth/token` | Back-channel call from your application server. |
| Authorization requests | `/authorize` | The application constructs this URL and triggers the redirect. Although the browser executes the request, the application controls the rate. |
| Device code requests | `/oauth/device/code` | Back-channel call to start device authorization. |
| Passwordless start | `/passwordless/start` | Back-channel call to initiate passwordless login. |
| Passwordless verify | `/passwordless/verify` | Back-channel call to verify a passwordless code. |
| Pushed authorization requests | `/oauth/par` | Back-channel call to push authorization parameters. |
| Token revocation | `/oauth/revoke` | Back-channel call to revoke a token. |
| Back-channel authorization (CIBA) | `/bc-authorize` | Back-channel call to initiate a CIBA authorization flow. |
| Cross-origin authentication | `/co/authenticate` | Back-channel call for cross-origin authentication. |
| Passkey challenge | `/passkey/challenge` | Back-channel call to initiate a passkey authentication challenge. |
| Passkey registration | `/passkey/register` | Back-channel call to register a new passkey. |

> ## Requests excluded from custom rate limits

Reviewer: ```suggestion
## Custom rate limit exclusions
```

Once Auth0 begins an authentication flow (after `/authorize`), the end user's browser interacts directly with Auth0 to complete login. These subsequent requests (rendering the login page, submitting credentials, completing MFA challenges) are not counted against custom rate limits. The application does not initiate these requests and cannot control their rate or timing.

For example, in a typical login flow:

1. Your application redirects the user to `/authorize` → **counted** (your app initiated this).
2. Auth0 renders the login page → **not counted** (browser-to-Auth0, outside your app's control).
3. The user submits their credentials → **not counted**.
4. The user completes an MFA challenge → **not counted**.
5. Auth0 redirects back with an authorization code → **not counted**.
6. Your application exchanges the code at `/oauth/token` → **counted** (your app initiated this).

<Callout icon="file-lines" color="#0EA5E9" iconType="regular">

> Because browser-initiated requests are excluded, custom rate limit policies reduce but do not eliminate the risk of attacks that exhaust tenant-wide rate limits by triggering complete authentication flows. For protection against those attack patterns, see [Attack Protection](/docs/secure/attack-protection).

Reviewer: ```suggestion
Because browser-initiated requests are excluded, custom rate limit policies reduce but do not eliminate the risk of attacks that exhaust tenant-wide rate limits by triggering complete authentication flows. For protection against those attack patterns, read [Attack Protection](/docs/secure/attack-protection).
```
